blackjax/vi/mfvi.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2020- The Blackjax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Callable, NamedTuple, Tuple, Union
+from blackjax.types import Array, PyTree
+from typing import NamedTuple
+
+import jax
+import jax.numpy as jnp
+import jax.random
+from jax.scipy import stats
+from jax import vmap
+
+
+__all__ = ["approximate", "sample", "mfvi_logprob", "MFVIState"]
+
+
+class MFVIState(NamedTuple):
+    elbo: Array
+    position: PyTree
+
+
+def mfvi_logprob(meanfield_param):
+    mu_param, rho_param = meanfield_param
+    sigma_param = jax.tree_map(jnp.exp, rho_param)
+
+    def meanfield_logprob(position):
+        logq_pytree = jax.tree_map(stats.norm.logpdf, position, mu_param, sigma_param)
+        logq = jax.tree_map(jnp.sum, logq_pytree)
+        return jax.tree_util.tree_reduce(jnp.add, logq)
+
+    return meanfield_logprob
+
+
+def sample(rng_key, meanfield_param, num_samples: int):
```

Comment:
Maybe you can call https://github.com/blackjax-devs/blackjax/blob/d6801f8a74721f881b5699bd536eea863e2a0009/blackjax/util.py#L56 internally?

Reply:
Sounds good.
But at this moment there's no internal function we can call to sample from multivariate Gaussian right (i.e. the fullrank VI case)?

Reply:
You can use the same function for multivariate Gaussian because the `linear_map` util will dispatch it correctly.

Reply:
@junpenglao But how do you generate num_sample of particles using `generate_gaussian_noise`

Reply:
You are right the util function does not accept that yet - let's keep your version here but could you add a TODO?
```
# TODO: switch to using `generate_gaussian_noise` in util.py
```